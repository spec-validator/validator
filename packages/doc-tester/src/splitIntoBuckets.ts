type Meta = {
  type: string,
  label: string
}

type CodeBlock = {
  lineno: number,
  lines: string[]
} & Meta

const hasDelimiter = (line: string): boolean => line.trim().startsWith('```')

export const extractType = (line: string): Meta => {
  const groups = line.match(/```(?<type>[^\s]+)?(\s+(?<label>[^\s]+))?/)?.groups

  const extract = (key: string): string =>
    (groups && groups[key]?.trim()?.toLowerCase()) || ''

  return {
    type: extract('type'),
    label: extract('label'),
  }
}

/**
 * Really primitive FSM based README parser aiming to extract code snippets
 * from the documentation.
 */
const extractCodeBlocks = (lines: string[], types: string[]): CodeBlock[] => {
  let block: CodeBlock | undefined = undefined
  const typeSet = new Set(types)
  const result: CodeBlock[] = []

  // eslint-disable-next-line max-statements
  lines.forEach((line, i) => {
    if (block) {
      //block.label = block.label || getLabelCheckpoint(line)
      if (hasDelimiter(line)) {
        if ((!typeSet.size || typeSet.has(block.type)) && !block.label.startsWith('#')) {
          result.push(block)
        }
        block = undefined
      } else {
        block.lines.push(line)
      }
    } else {
      if (hasDelimiter(line)) {
        block = {
          ...extractType(line),
          lineno: i + 1, // skip the delimiter - block - next line
          lines: [],
        }
      } else {
        // IGNORE
      }
    }
  })
  return result
}

export type Buckets = Record<string, CodeBlock[]>

const splitIntoBuckets = (blocks: CodeBlock[]): Buckets => {
  const buckets: Buckets = {}
  blocks.forEach(it => {
    const key = `${it.label}.${it.type}`
    buckets[key] = buckets[key] || []
    buckets[key].push(it)
  })
  return buckets
}

export const toCode = (blocks: CodeBlock[]): string => {
  const lines: string[] = []
  blocks.forEach(block => {
    while (lines.length < block.lineno) {
      lines.push('\n')
    }
    block.lines.forEach(it => {
      lines.push(it + '\n')
    })
  })
  return lines.join('')
}

export default (lines: string[], types: string[]=[]): Buckets =>
  splitIntoBuckets(extractCodeBlocks(lines, types))

