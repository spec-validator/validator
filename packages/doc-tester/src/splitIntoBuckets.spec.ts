
import splitIntoBuckets from './splitIntoBuckets'

test('untyped and unlabled blocks', () => {
  expect(splitIntoBuckets(
    [
      '# Title',
      '```',
      'CODE',
      '```',
      'Text',
      '```',
      'CODE 2',
      '```',
    ],
    ['']
  )).toMatchSnapshot()
})

test('typed and labled blocks', () => {
  expect(splitIntoBuckets(
    [
      '# Title',
      '```ts one',
      'CODE',
      '```',
      'Text',
      '```ts one',
      'CODE 2',
      '```',
    ],
    ['ts']
  )).toMatchSnapshot()
})
