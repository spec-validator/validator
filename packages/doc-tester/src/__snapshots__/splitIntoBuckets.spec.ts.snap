// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mixed blocks 1`] = `
Object {
  ".": Array [
    Object {
      "label": "",
      "lines": Array [
        "Unlabeled wildcard code",
      ],
      "offset": 15,
      "type": "",
    },
  ],
  ".ts": Array [
    Object {
      "label": "",
      "lines": Array [
        "Unlabeled ts code",
      ],
      "offset": 7,
      "type": "ts",
    },
  ],
  "one.": Array [
    Object {
      "label": "one",
      "lines": Array [
        "Labeled wildcard code",
      ],
      "offset": 11,
      "type": "",
    },
  ],
  "one.ts": Array [
    Object {
      "label": "one",
      "lines": Array [
        "Labeled ts code",
      ],
      "offset": 3,
      "type": "ts",
    },
  ],
}
`;

exports[`typed and labeled blocks 1`] = `
Object {
  "one.ts": Array [
    Object {
      "label": "one",
      "lines": Array [
        "CODE",
      ],
      "offset": 3,
      "type": "ts",
    },
    Object {
      "label": "one",
      "lines": Array [
        "CODE 2",
      ],
      "offset": 7,
      "type": "ts",
    },
  ],
}
`;

exports[`typed and unlabeled blocks 1`] = `
Object {
  ".ts": Array [
    Object {
      "label": "",
      "lines": Array [
        "CODE",
      ],
      "offset": 3,
      "type": "ts",
    },
    Object {
      "label": "",
      "lines": Array [
        "CODE 2",
      ],
      "offset": 7,
      "type": "ts",
    },
  ],
}
`;

exports[`untyped and labeled blocks 1`] = `
Object {
  "one.": Array [
    Object {
      "label": "one",
      "lines": Array [
        "CODE",
      ],
      "offset": 3,
      "type": "",
    },
    Object {
      "label": "one",
      "lines": Array [
        "CODE 2",
      ],
      "offset": 7,
      "type": "",
    },
  ],
}
`;

exports[`untyped and unlabeled blocks 1`] = `
Object {
  ".": Array [
    Object {
      "label": "",
      "lines": Array [
        "CODE",
      ],
      "offset": 3,
      "type": "",
    },
    Object {
      "label": "",
      "lines": Array [
        "CODE 2",
      ],
      "offset": 7,
      "type": "",
    },
  ],
}
`;
